vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs

```diff
@@ -0,0 +1,110 @@
+            let line = sourceText.Lines.[sourceText.Lines.IndexOf position]
+                
+            let defines = CompilerEnvironment.GetCompilationDefinesForEditing(document.FilePath, options.OtherOptions |> List.ofArray)
+
+            let tokens = Tokenizer.tokenizeLine(document.Id, sourceText, line.Start, document.FilePath, defines)
+
+            let! firstMeaningfulToken = 
+                tokens
+                |> List.tryFind (fun x ->
+                    x.Tag <> FSharpTokenTag.WHITESPACE &&
+                    x.Tag <> FSharpTokenTag.COMMENT &&
+                    x.Tag <> FSharpTokenTag.LINE_COMMENT)
+
+            let! (left, right) =
+                FSharpBraceMatchingService.GetBraceMatchingResult(checker, sourceText, document.FilePath, options, position, "FormattingService")
+
+            if right.StartColumn = firstMeaningfulToken.LeftColumn then
+                // Replace the indentation on this line with the indentation of the left bracket
+                let! leftSpan = RoslynHelpers.TryFSharpRangeToTextSpan(sourceText, left)
+
+                let indentChars (line : TextLine) =
+                    line.ToString()
+                    |> Seq.takeWhile ((=) ' ')
+                    |> Seq.length
+                    
+                let startIndent = indentChars sourceText.Lines.[sourceText.Lines.IndexOf leftSpan.Start]
+                let currentIndent = indentChars line
+
+                return TextChange(TextSpan(line.Start, currentIndent), String.replicate startIndent " ")
+            else
+                return! None
+        }
+
+    member __.GetFormattingChangesAsync (document: Document, position: int, cancellationToken: CancellationToken) =
+        async {
+            let! sourceText = document.GetTextAsync(cancellationToken) |> Async.AwaitTask
+            let optionsOpt = projectInfoManager.TryGetOptionsForEditingDocumentOrProject document
+            let! textChange = FSharpEditorFormattingService.GetFormattingChanges(document, sourceText, checkerProvider.Checker, optionsOpt, position)
+                
+            return
+                match textChange with
+                | Some change ->
+                    List([change]) :> IList<_>
+                | None ->
+                    List() :> IList<_>
+        }
+        
+    interface IEditorFormattingService with
+        member val SupportsFormatDocument = false
+        member val SupportsFormatSelection = false
+        member val SupportsFormatOnPaste = false
+        member val SupportsFormatOnReturn = true
+
+        override __.SupportsFormattingOnTypedCharacter (document, ch) =
+            if FSharpIndentationService.IsSmartIndentEnabled document.Project.Solution.Workspace then
+                match ch with
+                | ')' | ']' | '}' -> true
+                | _ -> false
+            else
+                false
+
+        override __.GetFormattingChangesAsync (_document, _span, _cancellationToken) = null
```

Comment:
The use of ``null`` in interface implementations always makes me a little nervous, could you add a comment indicating why this is valid, perhaps with a reference to a line in the Roslyn source that checks the null return value and does something sensible with it?